keps/sig-storage/20181111-extend-datasource-field.md

```diff
@@ -0,0 +1,111 @@
+
+### Goals
+
+* Remove restrictions imposed by API validation on Object types permitted to be set on the DataSource field
+* Enable ability to request a Clone of an existing PVC
+
+
+### Non-Goals
+
+* This does not verify the existence of the supplied Object Reference in the DataSource field
+* This does not provide any sort of action from Kubernetes on the object provided in the DataSource field
+
+## Proposal
+
+Remove Validation checks against the PVC DataSource field, due to the validation checks currently in place it's not possible to specify anything other than Snapshots as a DataSource.  Removing this check enables Cloning, as well as external object types.
+
+```yaml
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+    name: pvc-2
+    namespace: myns
+spec:
+  capacity:
+    storage: 10Gi
+  dataSource:
+    kind: PersistentVolumeClaim
+    name: pvc-1
+```
+
+The result would be a new and independent PV/PVC (pvc-2) on the backend device that is a duplicate of the data that existed on pvc-1.
+
+### User Stories [optional]
+
+#### Story 1
+As a cluster user, I want to be able to Clone a volume and run a different set of PODs/Applications against it
+
+### Implementation Details/Notes/Constraints [optional]
+
+The Kubernetes changes are minimal, and only require removal of the API Validation checks against allowed DataSource object types.  In order to maintain the existing behaviors, a feature gate "AnyDataSource" is added, with a efault enablement of False.  The existing SnapshotDataSource checks will still continue to function as expected if this is not enabled, if this feature is enabled however, any correctly formed API Group Kind can be specified as a DataSource.
+
+### Risks and Mitigations
+
+The primary risk identified with this proposal is that a DataSource object could be provided that doesn't exist on the Cluster.  In this case the PVC will be created but the user may be expecting some sort of action to take place based on the DataSource specifciation and that won't actually happen.
+
+## Graduation Criteria
+* API changes allowing specification of a PVC as a valid DataSource in a new PVC spec
+* Implementation of the PVC DataSource in the CSI external provisioner
+
+The API can be immediately promoted to Beta, as it's just an extension of the existing DataSource field.  There are no implementations or changes needed in Kubernetes other than accepting Object Types in the DataSource field.  This should be promoted to GA after a release assuming no major issues or changes were needed/required during the Beta stage.
+
+## Implementation History
+
+## Drawbacks [optional]
+
+## Alternatives [optional]
+
+Currently Cloning is implemented by most plugins through the use of annotations.  This however leads to an inconsistent user experience, and the annotation usage is not only informal but can be a bit unwieldly to use.
```

Comment:
Cloning can also be done by taking a snapshot and then create a PVC from the snapshot.

Reply:
That doesn't actually provide a mechanism to instruct a capable backend to do the clone; but yeah it is a work around.  I'll add a note, thanks!